### ISSUE TITLE
Using services in local development

### ISSUE THREAD

**Issue by trumbitta** [divshot/superstatic#132]
Today I needed to use the AJAX proxy service in my local development environment, but couldn't figure out how :)

I think it'd be nice to have a straightforward way for installing, or just using, superstatic-* services locally.

**Comment by scottcorgan** [comment]
We've been thinking through this for quite some in order to make the most elegant solution.

How do you feel about installing services locally through something like "superstatic install" or "divshot install" to install the services? Once installed, superstatic/divshot would know they're available and make them available.

**Comment by trumbitta** [comment]
Sounds perfect :+1: 

Also, maybe a slight variation like: 

```bash
$ [divshot | superstatic] service install
```
could lead to other useful subcommands like:

```bash
$ [divshot | superstatic] service [list | search | info]
```

**Comment by scottcorgan** [comment]
@trumbitta that's a very good suggestion.

@mbleigh what are your thoughts?

**Comment by scottcorgan** [comment]
Also, to keep consistent with the Divshot cli, I'm thinking: `superstatic services:install [name]`, etc.

**Comment by mbleigh** [comment]
How would that work under the covers? Local install or global?

**Comment by scottcorgan** [comment]
@mbleigh my though is that it's a global install into the `.divshot` folder in the home directory.

Another idea is that the user and use the `-g` flag or something like an `--local` flag in order to choose where it's locaed

(cc @trumbitta)

Thoughts?

**Comment by scottcorgan** [comment]
Note to self:

An idea I was thinking was to move services outside superstatic and let them be middleware that's easily implemented in a Express/Connect middleware chain. There seems to be no reason that services should only be tied to Superstatic. They solve good problems that can be useful outside Superstatic.

```js
var connect = require('connect');
var superstatic = require('superstatic');
var forms = require('superstatic-forms');

var app = connect();

app.use(forms(/* options */));
app.use(superstatic();

app.listen(3000, function () {});
```

**Comment by scottcorgan** [comment]
We've decided to read your superstatic.json/divshot.json file and search for superstatic services. If available, we'll check to see if you have them installed globally or locally. If you don't, then we'll automatically install them for you on the your next server start.

This was the easiest and clearest way to solve this problem and keep Superstatic simple.

This feature will be included in [Superstatic 2.1](https://github.com/divshot/superstatic/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.1) released next week.

**Comment by trumbitta** [comment]
:rocket: :shipit:

**Comment by scottcorgan** [comment]
Closed by https://github.com/divshot/superstatic/commit/9c479e846030d0ca419b349bf77e737de3a15be5.

Will publish today or tomorrow to NPM.

**Comment by dukedave** [comment]
@scottcorgan any update on when 'next week' will be? ;)
I still only see 2.0.2 [on npm](https://www.npmjs.com/package/superstatic).
Is there anyway to 'install' the proxy service in 2.0.2 in the interim?

Also, can I confirm that [this proxy config](http://docs.divshot.com/services/proxy) will 'just work' in 2.1?

**Comment by dukedave** [comment]
Hmm, it seems there was a mention of services in the README, but it was removed in 0fabe39dd0ff9bba06b621c7428088a9f6ec566d (I only thought to look because there's an orphaned link to `#services` at the top.

**Comment by scottcorgan** [comment]
@dukedave got sidetracked with trying to launch our number 1 requested feature on Divshot this week. Only left to call `2.1` is full ETag support and documentation for it.

Services have been in flux for a few months, that's why they're still in beta. That they work won't change, but how they work locally has been going back and forth. For now, if you `npm install <superstatic-service>` in the directory of your project and use the `--services proxy,forms` flag when starting superstatic, it will allow you to you use locally.

This will obviously be much easier as `2.1` is released with them automatically working if they're in your config file.

If you have any ideas on better ways to implement services locally, please let me know.

Thanks for your patience.

**Comment by dukedave** [comment]
Thanks for getting back @scottcorgan.
So, that makes perfect sense, except on 2.0.2 that `--services proxy` flag doesn't seem to have any effect (even with `superstatic-proxy` installed).
It works correctly when pushed to divshot, so I'm confident my `divshot.json` is correct.

The `services` flag is conspicuously absent from [cli/flags.js](https://github.com/divshot/superstatic/blob/master/lib/cli/flags.js), could that be it?

**Comment by scottcorgan** [comment]
You know what, you're right. I removed that flag in `2.x` because I knew it'd be changing. It was one of the breaking changes.

`divshot-cli` still uses superstatic `1.x`, so that flag should work with `divshot server --services proxy,forms`.

Try that and let me know.

**Comment by dukedave** [comment]
Rejoice, it works! Thanks again @scottcorgan :+1:

**Comment by scottcorgan** [comment]
@dukedave I promise `2.1` will be published soon. Ran into some hair-pullying snags on some bug fixes.

**Comment by scottcorgan** [comment]
@dukedave @trumbitta this has shipped in version `2.1.0`.

If you have configured the services in your divshot.json/superstatic.json file, it will automatically install and load!

**Comment by dukedave** [comment]
Excellent, thanks @scottcorgan
